--- a/init/Kconfig	2026-06-10 23:02:43.838376903 +0800
+++ a/init/Kconfig	2026-06-10 19:34:24.436437077 +0800
@@ -2270,3 +2270,10 @@
 # <asm/syscall_wrapper.h>.
 config ARCH_HAS_SYSCALL_WRAPPER
 	def_bool n
+
+config SOFTIRQ_THROTTLE
+	bool "Softirq Throttling Feature for Weak CPUs"
+	default y
+	help
+	  允许为软中断分配每 CPU 最大带宽。专门对付无 NAPI 导致的单核自旋。
+	  可以通过 /proc/sys/kernel/softirq_period_ms 和 softirq_runtime_ms 调节。
--- a/include/linux/interrupt.h	2024-03-27 06:23:12.000000000 +0800
+++ b/include/linux/interrupt.h	2026-06-10 19:32:38.274982413 +0800
@@ -565,7 +565,12 @@
 
 extern void raise_softirq_irqoff(unsigned int nr);
 extern void raise_softirq(unsigned int nr);
-
+#ifdef CONFIG_SOFTIRQ_THROTTLE
+extern unsigned int sysctl_softirq_period_ms;
+extern int sysctl_softirq_runtime_ms;
+int softirq_throttle_handler(struct ctl_table *table, int write, void *buffer,
+			     size_t *lenp, loff_t *ppos);
+#endif
 DECLARE_PER_CPU(struct task_struct *, ksoftirqd);
 
 static inline struct task_struct *this_cpu_ksoftirqd(void)
--- a/kernel/softirq.c	2024-03-27 06:23:12.000000000 +0800
+++ b/kernel/softirq.c	2026-06-13 15:18:56.095198640 +0800
@@ -61,6 +61,183 @@
 	"TASKLET", "SCHED", "HRTIMER", "RCU"
 };
 
+#ifdef CONFIG_SOFTIRQ_THROTTLE
+unsigned int sysctl_softirq_period_ms = 1000;
+int sysctl_softirq_runtime_ms = 900;
+unsigned int sysctl_softirq_block_ms = 10;
+unsigned int sysctl_softirq_throttle_mask = 3; // BIT(0) 0x1: NET_TX_SOFTIRQ | BIT(1) 0x2: NET_RX_SOFTIRQ
+
+struct softirq_stats {
+	u64 block_cnt;
+	u64 free_cnt;
+        u64 rx_block_cnt;
+        u64 tx_block_cnt;
+	u64 rcu_skip_cnt;
+};
+
+struct softnet_throttle_data {
+	u64 period;
+	u64 runtime;
+	raw_spinlock_t lock;
+} si_throttle;
+
+static DEFINE_PER_CPU(u64, si_time_start);
+static DEFINE_PER_CPU(u64, si_time_used);
+DEFINE_PER_CPU(u64, si_blocked_until);
+static DEFINE_PER_CPU(struct softirq_stats, softirq_stats_data);
+
+static int softirq_throttle_validate(void)
+{
+	if (((int)sysctl_softirq_period_ms <= 0) ||
+	    (sysctl_softirq_block_ms <= 0) ||
+	      (sysctl_softirq_block_ms > sysctl_softirq_period_ms) ||
+	    ((sysctl_softirq_runtime_ms != -1) &&
+	     ((unsigned int)sysctl_softirq_runtime_ms > sysctl_softirq_period_ms)) ||
+	    (sysctl_softirq_throttle_mask > 3))
+		return -EINVAL;
+	return 0;
+}
+
+static void softirq_throttle_update(void)
+{
+	unsigned long flags;
+	raw_spin_lock_irqsave(&si_throttle.lock, flags);
+	si_throttle.period = (u64)sysctl_softirq_period_ms * 1000000ULL;
+	si_throttle.runtime = (sysctl_softirq_runtime_ms == -1) ?
+		(u64)-1LL : (u64)sysctl_softirq_runtime_ms * 1000000ULL;
+	raw_spin_unlock_irqrestore(&si_throttle.lock, flags);
+}
+
+int softirq_throttle_handler(struct ctl_table *table, int write, void *buffer,
+			     size_t *lenp, loff_t *ppos)
+{
+	unsigned int old_period, old_runtime, old_block, old_mask;
+	static DEFINE_MUTEX(mutex);
+	int ret;
+
+	mutex_lock(&mutex);
+	old_period = sysctl_softirq_period_ms;
+	old_runtime = sysctl_softirq_runtime_ms;
+	old_block = sysctl_softirq_block_ms;
+	old_mask = sysctl_softirq_throttle_mask;
+
+	ret = proc_dointvec(table, write, buffer, lenp, ppos);
+	if (ret) goto undo;
+	if (!write) goto done;
+
+	ret = softirq_throttle_validate();
+	if (ret) goto undo;
+
+	softirq_throttle_update();
+	goto done;
+
+undo:
+	sysctl_softirq_period_ms = old_period;
+	sysctl_softirq_runtime_ms = old_runtime;
+	sysctl_softirq_block_ms = old_block;
+	sysctl_softirq_throttle_mask = old_mask;
+done:
+	mutex_unlock(&mutex);
+	return ret;
+}
+
+int sysctl_softirq_stats_handler(struct ctl_table *table, int write, void *buffer,
+				 size_t *lenp, loff_t *ppos)
+{
+	static DEFINE_MUTEX(stats_mutex);
+	int cpu; int ret;
+	char *buf;
+	int total_len = 1024;
+
+	if (write)
+		return -EPERM;
+
+	mutex_lock(&stats_mutex);
+	buf = kmalloc(total_len, GFP_KERNEL);
+	if (!buf) {
+		mutex_unlock(&stats_mutex);
+		return -ENOMEM;
+	}
+#define COLOR_GREEN	"\033[32m"
+#define COLOR_YELLOW	"\033[33m"
+#define COLOR_BLUE	"\033[36m"
+#define COLOR_RESTORE	"\033[0m"
+#define COLOR_WHITE	"\033[1m"
+
+	snprintf(buf, total_len, 
+		 COLOR_BLUE "=== Softirq Throttle Configuration ===\n" COLOR_RESTORE
+		 "throttle_mask *BIT(0) NET_TX_SOFTIRQ / *BIT(1) NET_RX_SOFTIRQ : " COLOR_WHITE "%u\n" COLOR_RESTORE
+		 "period_ms  : " COLOR_WHITE "%u ms\n" COLOR_RESTORE
+		 "runtime_ms : " COLOR_WHITE "%d ms\n" COLOR_RESTORE
+		 "block_ms   : " COLOR_WHITE "%u ms\n\n" COLOR_RESTORE
+		 COLOR_BLUE "=== Per-CPU Live Telemetry Stats ===\n" COLOR_RESTORE
+		 "cpu_id    blocked_count    freed_count    rx_blocked    tx_blocked    rcu_skipped\n",
+		 sysctl_softirq_throttle_mask,
+		 sysctl_softirq_period_ms, 
+		 sysctl_softirq_runtime_ms, 
+		 sysctl_softirq_block_ms);
+
+	for_each_possible_cpu(cpu) {
+		u64 b_cnt = per_cpu(softirq_stats_data, cpu).block_cnt;
+		u64 f_cnt = per_cpu(softirq_stats_data, cpu).free_cnt;
+		u64 rx_cnt = per_cpu(softirq_stats_data, cpu).rx_block_cnt;
+		u64 tx_cnt = per_cpu(softirq_stats_data, cpu).tx_block_cnt;
+		u64 rcu_cnt = per_cpu(softirq_stats_data, cpu).rcu_skip_cnt;
+
+		snprintf(buf + strlen(buf), total_len - strlen(buf), "cpu %d:    " COLOR_YELLOW "%-20llu" COLOR_GREEN "  %-15llu" COLOR_YELLOW "%-14llu" "%-14llu" COLOR_GREEN "%-15llu" COLOR_RESTORE "\n",
+			       cpu, b_cnt, f_cnt, rx_cnt, tx_cnt, rcu_cnt);
+	}
+
+	table->data = buf;
+	table->maxlen = strlen(buf) + 1;
+	ret = proc_dostring(table, write, buffer, lenp, ppos);
+
+	kfree(buf);
+	mutex_unlock(&stats_mutex);
+
+	return ret;
+}
+
+static void softirq_throttle_init(void)
+{
+	raw_spin_lock_init(&si_throttle.lock);
+	softirq_throttle_update();
+}
+
+static inline int check_softirq_throttled(u64 start_time)
+{
+	u64 now = ktime_get_mono_fast_ns();
+	u64 p_start = __this_cpu_read(si_time_start);
+	u64 used = __this_cpu_read(si_time_used);
+	u64 blocked = __this_cpu_read(si_blocked_until);
+	u64 chunk = now - start_time;
+
+	if (blocked && now < blocked) {
+		return 1;
+	} else if (blocked) {
+		__this_cpu_write(si_blocked_until, 0);
+		__this_cpu_inc(softirq_stats_data.free_cnt);
+	}
+
+	used += chunk;
+
+        if (si_throttle.runtime != (u64)-1LL && used >= si_throttle.runtime) {
+                __this_cpu_write(si_blocked_until, now + (u64)sysctl_softirq_block_ms * 1000000ULL);
+                __this_cpu_inc(softirq_stats_data.block_cnt);
+                return 1;
+        }
+
+        if (now - p_start >= si_throttle.period) {
+                __this_cpu_write(si_time_start, now);
+                __this_cpu_write(si_time_used, chunk);
+                return 0;
+        }
+
+	__this_cpu_write(si_time_used, used);
+	return 0;
+}
+#endif
+
 /*
  * we cannot loop indefinitely here to avoid userspace starvation,
  * but we also don't want to introduce a worst case 1/HZ latency
@@ -256,6 +433,11 @@
 	__u32 pending;
 	int softirq_bit;
 
+	/* patched 2026 */
+	u64 start_ns = ktime_get_mono_fast_ns();
+	max_restart = 10;
+	/* patched 2026 */
+
 	/*
 	 * Mask out PF_MEMALLOC as the current task context is borrowed for the
 	 * softirq. A softirq handled, such as network RX, might set PF_MEMALLOC
@@ -307,13 +489,41 @@
 
 	pending = local_softirq_pending();
 	if (pending) {
-		if (time_before(jiffies, end) && !need_resched() &&
-		    --max_restart)
+		if (ktime_get_mono_fast_ns() - start_ns < 2000000 && !need_resched() && --max_restart)
 			goto restart;
 
+                #ifdef CONFIG_SOFTIRQ_THROTTLE
+                if (check_softirq_throttled(start_ns)) {
+			if (pending & (1 << RCU_SOFTIRQ)) {
+				__this_cpu_inc(softirq_stats_data.rcu_skip_cnt);
+			} else {
+				unsigned int net_mask = 0;
+				if (sysctl_softirq_throttle_mask & BIT(0))
+					net_mask |= (1 << NET_TX_SOFTIRQ);
+				if (sysctl_softirq_throttle_mask & BIT(1))
+					net_mask |= (1 << NET_RX_SOFTIRQ);
+
+				if (net_mask && (pending & net_mask)) {
+					// would CFS be isolating this cpu by the kstat.
+					if (pending & (1 << NET_RX_SOFTIRQ)) {
+						kstat_incr_softirqs_this_cpu(NET_RX_SOFTIRQ);
+						__this_cpu_inc(softirq_stats_data.rx_block_cnt);
+					}
+					if (pending & (1 << NET_TX_SOFTIRQ)) {
+						kstat_incr_softirqs_this_cpu(NET_TX_SOFTIRQ);
+						__this_cpu_inc(softirq_stats_data.tx_block_cnt);
+					}
+
+					goto skip_restart;
+				}
+			}
+                }
+                #endif
+
 		wakeup_softirqd();
 	}
 
+skip_restart:
 	lockdep_softirq_end(in_hardirq);
 	account_irq_exit_time(current);
 	__local_bh_enable(SOFTIRQ_OFFSET);
@@ -572,10 +782,63 @@
 }
 EXPORT_SYMBOL(tasklet_kill);
 
+#ifdef CONFIG_SOFTIRQ_THROTTLE
+static struct ctl_table softirq_throttle_table[] = {
+	{
+		.procname   = "softirq_period_ms",
+		.data       = &sysctl_softirq_period_ms,
+		.maxlen     = sizeof(unsigned int),
+		.mode       = 0644,
+		.proc_handler   = softirq_throttle_handler,
+	},
+	{
+		.procname   = "softirq_runtime_ms",
+		.data       = &sysctl_softirq_runtime_ms,
+		.maxlen     = sizeof(int),
+		.mode       = 0644,
+		.proc_handler   = softirq_throttle_handler,
+	},
+	{
+		.procname   = "softirq_block_ms",
+		.data       = &sysctl_softirq_block_ms,
+		.maxlen     = sizeof(unsigned int),
+		.mode       = 0644,
+		.proc_handler   = softirq_throttle_handler,
+	},
+	{
+		.procname   = "softirq_throttle_mask",
+		.data       = &sysctl_softirq_throttle_mask,
+		.maxlen     = sizeof(unsigned int),
+		.mode       = 0644, 
+		.proc_handler   = softirq_throttle_handler,
+	},
+	{
+		.procname   = "softirq_stats",
+		.data       = NULL,
+		.maxlen     = 0,
+		.mode       = 0444, 
+		.proc_handler   = sysctl_softirq_stats_handler,
+	},
+	{ }
+};
+
+static __init int softirq_throttle_sysctl_init(void)
+{
+	register_sysctl_init("kernel", softirq_throttle_table);
+	return 0;
+}
+// 注册到内核早期的 initcall 钩子中
+early_initcall(softirq_throttle_sysctl_init);
+#endif
+
 void __init softirq_init(void)
 {
 	int cpu;
 
+	#ifdef CONFIG_SOFTIRQ_THROTTLE
+	softirq_throttle_init();
+	#endif
+
 	for_each_possible_cpu(cpu) {
 		per_cpu(tasklet_vec, cpu).tail =
 			&per_cpu(tasklet_vec, cpu).head;
@@ -592,8 +855,19 @@
 	return local_softirq_pending();
 }
 
+/* patched 2026 */
+#include <linux/netdevice.h>
+extern int netdev_max_backlog;
+DECLARE_PER_CPU(struct softnet_data, softnet_data);
+static DEFINE_PER_CPU(unsigned int, ksoftirq_yield_streak);
+/* patched 2026 */
 static void run_ksoftirqd(unsigned int cpu)
 {
+        #ifdef CONFIG_NET
+        unsigned int qlen;
+	unsigned int streak;
+        #endif
+	
 	local_irq_disable();
 	if (local_softirq_pending()) {
 		/*
@@ -601,6 +875,28 @@
 		 * in the task stack here.
 		 */
 		__do_softirq();
+		/* patched 2026 */
+		#ifdef CONFIG_NET
+		/*qlen = __this_cpu_read(softnet_data.input_pkt_queue.qlen);
+		if (qlen > (netdev_max_backlog >> 1)) { // 1/2 max_backlog
+			streak = __this_cpu_read(ksoftirq_yield_streak) + 1;
+			if (streak >= 3) {
+				__this_cpu_write(ksoftirq_yield_streak, 0);
+				local_irq_enable();
+				yield();
+				return;
+			} else {
+				__this_cpu_write(ksoftirq_yield_streak, streak);
+				local_irq_enable();
+				set_tsk_need_resched(current);
+				return;
+			}
+		} else {
+			if (__this_cpu_read(ksoftirq_yield_streak) > 0)
+				__this_cpu_write(ksoftirq_yield_streak, 0);
+		}*/
+		#endif
+		/* patched 2026 */
 		local_irq_enable();
 		cond_resched();
 		return;
--- a/net/core/dev.c	2026-06-14 05:43:07.863400432 +0800
+++ b/net/core/dev.c	2026-06-14 05:41:43.339013398 +0800
@@ -4441,6 +4441,241 @@
 }
 EXPORT_SYMBOL_GPL(do_xdp_generic);
 
+#ifdef CONFIG_SOFTIRQ_THROTTLE
+static int copy_of_netif_rx_internal(struct sk_buff *skb)
+{
+        int ret;
+
+        net_timestamp_check(READ_ONCE(netdev_tstamp_prequeue), skb);
+
+        trace_netif_rx(skb);
+
+#ifdef CONFIG_RPS
+        if (static_branch_unlikely(&rps_needed)) {
+                struct rps_dev_flow voidflow, *rflow = &voidflow;
+                int cpu;
+
+                preempt_disable();
+                rcu_read_lock();
+
+                cpu = get_rps_cpu(skb->dev, skb, &rflow);
+                if (cpu < 0)
+                        cpu = smp_processor_id();
+
+                ret = enqueue_to_backlog(skb, cpu, &rflow->last_qtail);
+
+                rcu_read_unlock();
+                preempt_enable();
+        } else
+#endif
+        {
+                unsigned int qtail;
+
+                ret = enqueue_to_backlog(skb, get_cpu(), &qtail);
+                put_cpu();
+        }
+        return ret;
+}
+
+static DEFINE_PER_CPU(struct sk_buff_head, mtk_batch_queue);
+static DEFINE_PER_CPU(u8, mtk_batch_count) = 0;
+static DEFINE_PER_CPU(u64, last_flush_ns) = 0;
+
+int sysctl_rx_enable_batch = 0;
+unsigned int sysctl_rx_batch_us = 10;
+unsigned int sysctl_rx_max_batch_cnt = 4;
+DEFINE_PER_CPU(u64, rx_success_batch_sum) = 0;
+
+static int __netif_rx_internal(struct sk_buff *skb, bool flush)
+{
+	int ret = NET_RX_SUCCESS;
+	unsigned int batch_cnt;
+	struct sk_buff_head *batch_q;
+	u64 now_ns;
+
+	if (skb) {
+		net_timestamp_check(READ_ONCE(netdev_tstamp_prequeue), skb);
+	}
+
+#ifdef CONFIG_RPS
+	preempt_disable();
+#endif
+	batch_q = this_cpu_ptr(&mtk_batch_queue);
+
+	if (unlikely(!batch_q->next)) {
+		skb_queue_head_init(batch_q);
+	}
+
+	if (skb) {
+		__skb_queue_tail(batch_q, skb);
+		__this_cpu_inc(mtk_batch_count);
+	}
+	batch_cnt = __this_cpu_read(mtk_batch_count);
+
+	now_ns = ktime_get_mono_fast_ns();
+
+	if (!flush && likely(batch_cnt < sysctl_rx_max_batch_cnt && 
+		(now_ns - __this_cpu_read(last_flush_ns)) < sysctl_rx_batch_us*1000)) {
+#ifdef CONFIG_RPS
+		preempt_enable();
+#endif
+		return NET_RX_SUCCESS;
+	}
+
+#ifdef CONFIG_RPS
+	rcu_read_lock();
+#endif
+	__this_cpu_write(last_flush_ns, now_ns);
+
+	while ((skb = __skb_dequeue(batch_q))) {
+#ifdef CONFIG_RPS
+		if (static_branch_unlikely(&rps_needed)) {
+			struct rps_dev_flow voidflow, *rflow = &voidflow;
+			int cpu;
+
+			cpu = get_rps_cpu(skb->dev, skb, &rflow);
+			if (cpu < 0)
+				cpu = smp_processor_id();
+			ret = enqueue_to_backlog(skb, cpu, &rflow->last_qtail);
+		} else
+#endif
+		{
+			unsigned int qtail;
+			ret = enqueue_to_backlog(skb, smp_processor_id(), &qtail);
+			put_cpu();
+		}
+	}
+
+	if (batch_cnt >= sysctl_rx_max_batch_cnt)
+		__this_cpu_inc(rx_success_batch_sum); // actual aggr num!
+	__this_cpu_write(mtk_batch_count, 0);
+
+#ifdef CONFIG_RPS
+	rcu_read_unlock();
+	preempt_enable();
+#endif
+	return ret;
+}
+
+static int netif_rx_internal(struct sk_buff *skb)
+{
+	if (!sysctl_rx_enable_batch ||
+		!skb || !skb->dev || (skb->dev->flags & IFF_LOOPBACK) || skb->dev->rtnl_link_ops)
+		return copy_of_netif_rx_internal(skb);
+
+	return __netif_rx_internal(skb, false); // flush = false
+}
+
+static void netif_rx_batch_flush_remote(void *info)
+{
+	unsigned long flags;
+	local_irq_save(flags);
+
+	__netif_rx_internal(NULL, true);
+
+	local_irq_restore(flags);
+}
+
+static struct timer_list rx_flush_timer;
+#define MAX_RX_BATCH_FLUSH_MS 10 // ms
+DEFINE_PER_CPU(struct __call_single_data, rx_flush_csd);
+static void netif_rx_flush_timer_fn(struct timer_list *t)
+{
+	int cpu;
+
+	for_each_online_cpu(cpu) {
+		if (per_cpu(mtk_batch_count, cpu) > 0) {
+			struct __call_single_data *csd = &per_cpu(rx_flush_csd, cpu);
+			smp_call_function_single_async(cpu, csd);
+		}
+	}
+
+	mod_timer(&rx_flush_timer, jiffies + msecs_to_jiffies(MAX_RX_BATCH_FLUSH_MS));
+}
+
+int sysctl_netif_rx_handler(struct ctl_table *table, int write, void *buffer,
+                                 size_t *lenp, loff_t *ppos)
+{
+        static DEFINE_MUTEX(mutex);
+        int cpu; int ret;
+        char *buf;
+        int total_len = 256;
+
+        if (write)
+                return -EPERM;
+
+        mutex_lock(&mutex);
+        buf = kmalloc(total_len, GFP_KERNEL);
+        if (!buf) {
+                mutex_unlock(&mutex);
+                return -ENOMEM;
+        }
+
+        snprintf(buf, total_len, "\n");
+	snprintf(buf + strlen(buf), total_len - strlen(buf), 
+			"NETIF RX BATCH CONFIG: \nrx_batch_us: %u\trx_max_batch_cnt: %hhu\n\n", 
+			sysctl_rx_batch_us, sysctl_rx_max_batch_cnt);
+
+        for_each_possible_cpu(cpu) {
+                u64 cnt = per_cpu(rx_success_batch_sum, cpu);
+
+                snprintf(buf + strlen(buf), total_len - strlen(buf), "cpu %d: %llu\t",
+                               cpu, cnt);
+        }
+
+        table->data = buf;
+        table->maxlen = strlen(buf) + 1;
+        ret = proc_dostring(table, write, buffer, lenp, ppos);
+
+        kfree(buf);
+        mutex_unlock(&mutex);
+
+        return ret;
+}
+
+static struct ctl_table netif_rx_batch_table[] = {
+        {
+                .procname   = "rx_batch_us",
+                .data       = &sysctl_rx_batch_us,
+                .maxlen     = sizeof(unsigned int),
+                .mode       = 0644,
+                .proc_handler = proc_dointvec,
+        },
+        {
+                .procname   = "rx_max_batch_pkts",
+                .data       = &sysctl_rx_max_batch_cnt,
+                .maxlen     = sizeof(unsigned int),
+                .mode       = 0644,
+		.proc_handler = proc_douintvec_minmax,
+		.extra1       = SYSCTL_ONE,
+		.extra2       = SYSCTL_INT_MAX,
+        },
+	{
+		.procname     = "rx_enable_batch",
+		.data         = &sysctl_rx_enable_batch,
+		.maxlen       = sizeof(int),
+		.mode         = 0644,
+		.proc_handler = proc_dointvec_minmax,
+		.extra1       = SYSCTL_ZERO,
+		.extra2       = SYSCTL_ONE,
+	},
+        {
+                .procname   = "rx_batch_stats",
+                .data       = NULL,
+                .maxlen     = 0,
+                .mode       = 0444,
+                .proc_handler = sysctl_netif_rx_handler,
+        },
+        { }
+};
+
+static __init int netif_rx_batch_sysctl_init(void)
+{
+        register_sysctl_init("net/core", netif_rx_batch_table);
+        return 0;
+}
+early_initcall(netif_rx_batch_sysctl_init);
+#else /* CONFIG_SOFTIRQ_THROTTLE */
 static int netif_rx_internal(struct sk_buff *skb)
 {
 	int ret;
@@ -4475,6 +4710,7 @@
 	}
 	return ret;
 }
+#endif
 
 /**
  *	netif_rx	-	post buffer to the network code
@@ -6546,7 +6782,11 @@
 	list_splice(&list, &sd->poll_list);
 	if (!list_empty(&sd->poll_list))
 		__raise_softirq_irqoff(NET_RX_SOFTIRQ);
-
+#ifdef CONFIG_SOFTIRQ_THROTTLE
+        if (__this_cpu_read(mtk_batch_count) > 0 && (ktime_get_mono_fast_ns() - __this_cpu_read(last_flush_ns)) >= (sysctl_rx_batch_us * 1000)) {
+                __netif_rx_internal(NULL, true);
+	}
+#endif
 	net_rps_action_and_irq_enable(sd);
 out:
 	__kfree_skb_flush();
@@ -10482,6 +10722,19 @@
 				       NULL, dev_cpu_dead);
 	WARN_ON(rc < 0);
 	rc = 0;
+
+#ifdef CONFIG_SOFTIRQ_THROTTLE
+	/*
+	for_each_possible_cpu(i) {
+		struct __call_single_data *csd = &per_cpu(rx_flush_csd, i);
+		
+		csd->func = netif_rx_batch_flush_remote; 
+		csd->info = NULL;                      
+	}
+	timer_setup(&rx_flush_timer, netif_rx_flush_timer_fn, 0);
+	mod_timer(&rx_flush_timer, jiffies + msecs_to_jiffies(MAX_RX_BATCH_FLUSH_MS));
+	*/
+#endif
 out:
 	return rc;
 }
